### PR TITLE
Vim mode

### PR DESCRIPTION
**Pull request by idealemu**
**Summary:**
Adding an editor Vim-like mode to novelWriter. This will enable users who activate this feature to navigate files using basic vim motions, performing basic editing in a vim "Normal mode" and write text normally in the "Insert mode".

**Related Issue(s):**
https://github.com/vkbo/novelWriter/issues/2493
**Reviewer's Checklist:**

* [ ] The header of all files contain a reference to the repository license
* [ ] The overall test coverage is increased or remains the same as before
* [ ] All tests are passing
* [ ] All linting checks are passing and the style guide is followed
* [ ] Documentation (as docstrings) is complete and understandable
* [ ] Only files that have been actively changed are committed

### CODE REVIEW

**Comment by vkbo** [comment]
Just an FYI (I already fixed it) under "Related Issue(s)" you put "Closes" and the issue number (not the link). It then triggers the automatic linking to the ticket.

File: novelwriter/gui/doceditor.py [hunk 1]
```diff
@@ -919,6 +934,10 @@ def keyPressEvent(self, event: QKeyEvent) -> None:
           * We also handle automatic scrolling here.
         """
         self._lastActive = time()
+
+        if self._handleVimKeyPress(event):
+            return
```
Comment (vkbo) [inline novelwriter/gui/doceditor.py:939]: Could we check `CONFIG.vimModeEnabled` here instead of inside the function so that we skip the function call if it isn't enabled? There is some overhead for calling functions and it's best to avoid it if it isn't necessary.

Something like:
```python
if CONFIG.vimModeEnabled and self._handleVimKeyPress(event):
    return
```

Comment (vkbo) [reply novelwriter/gui/doceditor.py:939]: Hmm, it seems the function always returns True if Vim Mode is enabled, so maybe the whole boolean return can be dropped and instead use:

```python
if CONFIG.vimModeEnabled:
    self._handleVimKeyPress(event)
    return
```

**Comment by vkbo** [comment]
This is looking really great. I'll see if I can fix my part of it this weekend.

Test coverage is still a little spotty. You can click forward to the test coverage report from the codecov check here on the pull request, or if you run tests locally with `run_tests.py`, setting the `-r` flag will generate a HTML report folder at the root of the repo. On Linux, the coverage report should show 26 missing lines when the code is covered 100%. The remaining lines are covered on Windows and MacOS.

**Comment by vkbo** [comment]
I'm also happy to write some of the tests. It helps to understand the code flow. I also find that writing tests is a good way to test the logic of the code. I often discover issues and corner cases when writing tests.

**Comment by idealemu** [comment]
> This is looking really great. I'll see if I can fix my part of it this weekend.
> 
> Test coverage is still a little spotty. You can click forward to the test coverage report from the codecov check here on the pull request, or if you run tests locally with `run_tests.py`, setting the `-r` flag will generate a HTML report folder at the root of the repo. On Linux, the coverage report should show 26 missing lines when the code is covered 100%. The remaining lines are covered on Windows and MacOS.

Super ! I did a trial usage of it yesterday evening and I have one or two features I realised I was missing. I'll bump the test coverage up once those are added. :smile:

**Comment by vkbo** [comment]
Not sure if you've noticed, but the codecov job links through to a report that shows what lines are not covered by tests. Only 7 lines are missed now!

https://app.codecov.io/gh/vkbo/novelWriter/pull/2499

**Comment by vkbo** [comment]
I fixed the isort linting complaint. It's a little clunky to set up because ruff and isort interfere a little with each other. Unfortunately, import sorting in ruff is a litt under-configurable.

**Comment by vkbo** [comment]
In any case, the PR looks complete to me in terms of features. I can do the parts I need to do in a follow up PR. It may be cleaner.

**Comment by idealemu** [comment]
> I fixed the isort linting complaint. It's a little clunky to set up because ruff and isort interfere a little with each other. Unfortunately, import sorting in ruff is a litt under-configurable.

Yes, I couldn't quite figure out what was wrong with it. Thanks !

> In any case, the PR looks complete to me in terms of features. I can do the parts I need to do in a follow up PR. It may be cleaner.

In my eyes almost ! I have two or three other basic commands I would like to add before calling it complete (and corresponding test coverage), and I would like to refactor everything a little, but I could do the refactor in another PR if you want ? It would just be refactor of the _handleVim method, and of the test cases which currently are really just "add everything until coverage is hit" without much organisation.

**Comment by vkbo** [comment]
> Yes, I couldn't quite figure out what was wrong with it. Thanks !

If you use VSCode, and have Microsoft's Isort extension installed, add this config to the `settings.json` file for the workspace or `.vscode` folder:

```json
    "[python]": {
        "editor.defaultFormatter": "ms-python.isort",
        "editor.codeActionsOnSave": {
            "source.organizeImports": "always",
            "source.fixAll": "always"
        },
    },
```

The needed settings are in `pyproject.toml`, so the additional config above should ensure that the imports are auto-fixed when you save.

You can also run:

```bash
isort --check .
```

I'm considering enabling Ruff formatting for the code base, but it lacks a few features I need, so I haven't done so yet.

> In my eyes almost ! I have two or three other basic commands I would like to add before calling it complete (and corresponding test coverage), and I would like to refactor everything a little, but I could do the refactor in another PR if you want ? It would just be refactor of the _handleVim method, and of the test cases which currently are really just "add everything until coverage is hit" without much organisation.

Sure, no rush.

**Comment by idealemu** [comment]
I have now added what I think are my final features, checked test code coverage, and refactored code and tests into something somewhat readable :smile: 
It is still quite if statement heavy but seeing this whole thing is essentially matching different user inputs to commands I think this cannot be avoided. Please let me know if anything seems out of style or wrong.

> I can do the parts I need to do in a follow up PR. It may be cleaner.

If by that you mean hooking the config into the settings then that's super fine by me :+1:

**Comment by vkbo** [comment]
Great! I'll do a review as soon as I have some extra time. I'm quite busy these days with work, so it may take a bit.

With a quick glanse at the code I can see I will have some optimisation suggestions here and there, but as far as I can see, the structure seems fine to me. Looks good!

File: novelwriter/gui/doceditor.py [hunk 2]
```diff
@@ -3220,3 +3585,74 @@ def updateMainCount(self, count: int, selection: bool) -> None:
...
+
+    def resetCommand(self) -> None:
+        """Reset internal vim command."""
+        self._normalCommand = ""
+        self._visualCommand = ""
+
+    def setMode(self, new_mode: nwVimMode) -> None:
+        """Switch vim mode."""
+        self._mode = new_mode
+        self.resetCommand()
```
Comment (vkbo) [inline novelwriter/gui/doceditor.py:3625]:
```suggestion
    def setMode(self, newMode: nwVimMode) -> None:
        """Switch vim mode."""
        self._mode = newMode
        self.resetCommand()
```

File: novelwriter/gui/doceditor.py [hunk 3]
```diff
@@ -3220,3 +3585,74 @@ def updateMainCount(self, count: int, selection: bool) -> None:
...
+        "_internalClipboard",
+        "_mode",
+        "_normalCommand",
+        "_visualCommand",
+    )
+
+    def __init__(self) -> None:
+        self.PREFIX_KEYS = ["d", "y", "g", "z"]
+        self.VISUAL_PREFIX_KEYS = ["g"]
+        self.SUFFIX_KEYS = ["w"]
```
Comment (vkbo) [inline novelwriter/gui/doceditor.py:3610]: As far as I can tell, these lists can just be changed to strings. A string is also classified as a sequence in Python. As long as you only check a single character, there's no need to make them lists.

```suggestion
        self.PREFIX_KEYS = "dygz"
        self.VISUAL_PREFIX_KEYS = "g"
        self.SUFFIX_KEYS = "w"
```

File: novelwriter/gui/doceditor.py [hunk 4]
```diff
@@ -3220,3 +3585,74 @@ def updateMainCount(self, count: int, selection: bool) -> None:
...
+        self._visualCommand = ""
+
+    def setMode(self, new_mode: nwVimMode) -> None:
+        """Switch vim mode."""
+        self._mode = new_mode
+        self.resetCommand()
+
+    def getMode(self) -> nwVimMode:
+        """Return current vim mode."""
+        return self._mode
```
Comment (vkbo) [inline novelwriter/gui/doceditor.py:3629]: Generally, I try to avoid get methods if they are plain and simple like this one and instead use Python property instead.

```suggestion
    @property
    def mode(self) -> nwVimMode:
        """Return current vim mode."""
        return self._mode
```

**Comment by vkbo** [comment]
I looked at the code, and it's mostly just nitpick. I can also fix this myself if you wish since none of it changes the functionality.

One more suggestion is to move the private methods used for Vim Mode to a separate section in the class. As you can see around the code, I tend to sort the methods into sections. Mainly they are:

1. Magic methods (`__init__`, `__bool__`, `__len__`, etc) 
2. Properties
3. Setter methods
4. General public functionality methods
5. Event handlers/overloads
6. Public slots
7. Private slots
8. Internal functions

Of course, the doc editor source code file is huge, so I have tried to organise them also by topic. Like all the search/replace methods in one section. I think your Vim private methods are big enough to warrant another such section, so I would like them moved to a section of their own above the "Internal Functions" section.

So basically, something like:

```python
    ##
    #  Internal Functions : Vim Mode
    ##

    def _handleVimNormalModeModeSwitching(self, event: QKeyEvent) -> bool:
        ...

    def _handleVimNormalMode(self, event: QKeyEvent) -> bool:
        ...

    def _handleVimVisualMode(self, event: QKeyEvent) -> bool:
        ...

    def _handleVimKeyPress(self, event: QKeyEvent) -> None:
        ...

    ##
    #  Internal Functions
    ##
```

File: novelwriter/gui/doceditor.py [hunk 5]
```diff
@@ -948,6 +968,337 @@ def keyPressEvent(self, event: QKeyEvent) -> None:
...
+            cursor = self.textCursor()
+            cursor.movePosition(cursor.MoveOperation.StartOfLine)
+            self.setTextCursor(cursor)
+            self.setVimMode(nwVimMode.INSERT)
+            return True
+        elif event.text() == "v":
+            self.setVimMode(nwVimMode.VISUAL)
+            cursor = self.textCursor()
+            self._visualAnchor = cursor.position()
+            cursor.movePosition(cursor.MoveOperation.Right, cursor.MoveMode.KeepAnchor, 1)
```
Comment (vkbo) [inline novelwriter/gui/doceditor.py:985]: These enums should be extracted from `QTextCursor` itself, not the instance of the cursor. It makes them easier to find later, and it's consistent with the rest of the code.

Also, most of the enums that are frequently accessed are defined with a short name variable in `types.py`. Many of the ones you use are already defined there. Feel free to add more if you find they are used often.

I added these shortnames when I ported novelWriter from PyQt5 to PyQt6. Since PyQt6 you have to use the full namespace path of the enum values, which means they tend to be very long, and since these values have to be looked up on each dot, is a bit of a waste of cycles.

I'm trying to be very conscious of performance on all code that is triggered on key press as Python is relatively slow, and bouncing calls between Python and the Qt library has a lot of overhead due to type conversion back and forth. There are already indications that some users are experiencing typing lag which may or may not be due to this. There are also typing lag issues related to OS rendering of the text, which is unrelated, but I'm trying to not add to the problem.

**Comment by vkbo** [comment]
~~Let me know if you want me to just fix these myself. I've been away a lot lately, so I haven't had time to work on novelWriter, but I hope to find some time soon.~~

**Comment by vkbo** [comment]
Hi again. I've updated the PR and fixed the bits I have mentioned and a bunch of other things for consistency with the rest of the code. It looks fine to me now. The `vimmode` variable is also now present in the `novelwriter.conf` file, but not added in Preferences yet. I'll do that in a separate PR later.,

However, a major problem for me is that I don't understand how Vim Mode actually works, so it's difficult for me to test that my changes didn't break anything. Could you give it a test before we merge it?

**Comment by vkbo** [comment]
I've looked through the vim command processing functions, and I think they can be simplified quite a bit as well as improved for performance. However, I don't want to do this in the current PR. When you've verified that my current changes don't break anything, I'll make a new PR with the refactoring and we can compare the two.

**Comment by vkbo** [comment]
I also added the colours from the discussion to the Vim mode labels on the editor footer. You can change them if you wish. I just picked the ones from the screenshot and added blue for insert so it stands out.

**Comment by vkbo** [comment]
Hi, I haven't heard from you in a while. I want to make the beta release very soon, but there are a couple of steps left before this can go in. I mainly just need you to do a functional test of the Vim mode after my changes to your PR. The tests pass, so I expect it works as it should.

I also want to do some optimisations in another PR, but I need an answer to the [remaining comment](https://github.com/vkbo/novelWriter/pull/2499#discussion_r2404079701) for my optimisation rewrite.

I also need documentation for the Vim mode. If you don't know how to write reStructuredText docs, you can post it as Markdown and I'll adapt it (just create `docs/vim_mode.md` or something). It can be done in a separate PR, so no rush for this one to be merged, but I need it before the RC1 release. I think a short introduction and a table listing the different modes and what codes are available and what they do should be enough. I assume most people who know what Vim mode is already know what the commands mean, but a quick overview is anyway helpful and also documents which commands are implemented.

**Comment by idealemu** [comment]
> Hi, I haven't heard from you in a while. I want to make the beta release very soon, but there are a couple of steps left before this can go in. I mainly just need you to do a functional test of the Vim mode after my changes to your PR. The tests pass, so I expect it works as it should.
> 
> I also want to do some optimisations in another PR, but I need an answer to the [remaining comment](https://github.com/vkbo/novelWriter/pull/2499#discussion_r2404079701) for my optimisation rewrite.
> 
> I also need documentation for the Vim mode. If you don't know how to write reStructuredText docs, you can post it as Markdown and I'll adapt it (just create `docs/vim_mode.md` or something). It can be done in a separate PR, so no rush for this one to be merged, but I need it before the RC1 release. I think a short introduction and a table listing the different modes and what codes are available and what they do should be enough. I assume most people who know what Vim mode is already know what the commands mean, but a quick overview is anyway helpful and also documents which commands are implemented.

Hi, sorry for the long gap. I will check up on the [remaining comment](https://github.com/vkbo/novelWriter/pull/2499#discussion_r2404079701) later today.

I will add the content for the docs as a markdown in `docs/vim_mode.md`. It will have an introduction, list of implemented commands, and any particular quirks that diverge a little from normal vim behaviour. 

I will also redo some functionality testing and check everything works as intended. Thank you again for the changes and taking the time to review !

**Comment by vkbo** [comment]
Great, thanks!

I have a performance optimisation branch as well where I only read the `self._vim.command` value once and cache it for the rest of the if-elif. As far as I can tell from the code, that shouldn't be a problem. The only exception from the pattern is the one elif that doesn't return `True`. If that's just something you forgot, my rewrite should be fine. (It passes your tests.)

File: novelwriter/gui/doceditor.py [hunk 6]
```diff
@@ -1879,6 +1909,328 @@ def _insertCommentStructure(self, style: nwComment) -> None:
...
+            return True
+
+        if self._vim.command == "O":
+            cursor.beginEditBlock()
+            cursor.movePosition(QtMoveStartOfLine)
+            cursor.insertText("\n")
+            cursor.movePosition(QtMoveUp)
+            cursor.endEditBlock()
+            self.setTextCursor(cursor)
+            self.setVimMode(nwVimMode.INSERT)
```
[the remark this answers is not in the document]
Comment (idealemu) [reply novelwriter/gui/doceditor.py]: I'm 99% sure this is just that I forgot it.


Currently because of the behaviour of `_handleVimKeyPress` if it would not have the return True it would just move to the next statement which is nothing in this case. 

```python
    def _handleVimKeyPress(self, event: QKeyEvent) -> None:
        """Handle key events for Vim mode.
        If vim mode is not enabled, typing behaves as normal.
        """
        # --- INSERT mode, bypass ---
        if self._vim.mode == nwVimMode.INSERT:
            super().keyPressEvent(event)
            return  # Normal typing

        if self._handleVimNormalModeModeSwitching(event):
            return

        if self._handleVimVisualMode(event):
            return

        if self._handleVimNormalMode(event):
            return

        return
```

Could be simplified to:
```python
    def _handleVimKeyPress(self, event: QKeyEvent) -> None:
        """Handle key events for Vim mode.
        If vim mode is not enabled, typing behaves as normal.
        """
        # --- INSERT mode, bypass ---
        if self._vim.mode == nwVimMode.INSERT:
            super().keyPressEvent(event)
            return  # Normal typing

        if self._handleVimNormalModeModeSwitching(event):
            return

        if self._handleVimVisualMode(event):
            return

        self._handleVimNormalMode(event)  # Don't care about return since it's the last call
```

But we would still want early return statements in the handleVim methods to stop evaluating all the if statements :)

**Comment by idealemu** [comment]
>The vimmode variable is also now present in the novelwriter.conf file

I can't seem to find a file with that exact name ? Could you point me towards how I can enable vim mode until it is in the preferences ? 
I can see `tests/reference/baseConfig_novelwriter.conf` but that would only enable it for the test right ? Is there another path to enableing the config and just 'free using' novelWriter ? 

I can still hardcode in `novelwriter/config.py`, so just wondering if there is a better way in place currently until the config hits the GUI.

**Comment by vkbo** [comment]
> I can't seem to find a file with that exact name ? 

See https://novelwriter.io/docs/technical/locations.html#configuration

**Comment by idealemu** [comment]
I have added `docs/vim-mode.md` with basic documentation.

I tested while doing some writing and everything seems to work fine.
However, I also noticed that there are two commands "db" and "de" are missing (delete back, and delete end, which are both analogous to delete word).

Based on existing code I would suggest:

```python

if self._vim.command == "db":
    cursor.beginEditBlock()
    cursor.movePosition(QtMovePreviousWord, QtKeepAnchor)
    self._vim.yankToInternal(cursor.selectedText())
    cursor.removeSelectedText()
    cursor.endEditBlock()
    self.setTextCursor(cursor)
    self._vim.resetCommand()
    return True

if self._vim.command == "de":
    cursor.beginEditBlock()
    # Extend selection to end of current/next word
    origPos = cursor.position()
    cursor.movePosition(QtMoveEndOfWord, QtKeepAnchor)
    if cursor.position() == origPos:               # already at end-of-word
        textLen = len(self.toPlainText())
        if origPos < textLen:
            cursor.movePosition(QtMoveNextChar, QtKeepAnchor)
            cursor.movePosition(QtMoveEndOfWord, QtKeepAnchor)
    self._vim.yankToInternal(cursor.selectedText())
    cursor.removeSelectedText()
    cursor.endEditBlock()
    self.setTextCursor(cursor)
    self._vim.resetCommand()
    return True
```
But I need to seem if this matches the performances changes you made :smiley: 
Otherwise I think I am decently satisfied with what this PR accomplishes.

**Comment by vkbo** [comment]
Don't worry about the performance changes. I can adapt those to fit.

I see there are merge conflict issues with the current main branch. I can fix those if you wish, once you're done with your changes.

**Comment by idealemu** [comment]
Super ! I just added these two, and corresponding test coverage and doc update then :) If you could take care of the conflicts that would be great ! Docs are currently markdown as you suggested, I wanted to learn the format you use but I frankly kind of just want this PR to be ready :smile:

**Comment by vkbo** [comment]
Updated your branch and fixed linting errors.

It looks good. One minor thing is that the "Already at end-of-word" condition here is not covered by tests.

```python
        if self._vim.command == "de":
            cursor.beginEditBlock()
            # Extend selection to end of current/next word
            origPos = cursor.position()
            cursor.movePosition(QtMoveEndOfWord, QtKeepAnchor)
            if cursor.position() == origPos:  # Already at end-of-word
                textLen = len(self.toPlainText())
                if origPos < textLen:
                    cursor.movePosition(QtMoveNextChar, QtKeepAnchor)
                    cursor.movePosition(QtMoveEndOfWord, QtKeepAnchor)
            self._vim.yankToInternal(cursor.selectedText())
            cursor.removeSelectedText()
            cursor.endEditBlock()
            self.setTextCursor(cursor)
            self._vim.resetCommand()
            return True
```

**Comment by idealemu** [comment]
> Updated your branch and fixed linting errors.
> 
> It looks good. One minor thing is that the "Already at end-of-word" condition here is not covered by tests.
> 
> ```python
>         if self._vim.command == "de":
>             cursor.beginEditBlock()
>             # Extend selection to end of current/next word
>             origPos = cursor.position()
>             cursor.movePosition(QtMoveEndOfWord, QtKeepAnchor)
>             if cursor.position() == origPos:  # Already at end-of-word
>                 textLen = len(self.toPlainText())
>                 if origPos < textLen:
>                     cursor.movePosition(QtMoveNextChar, QtKeepAnchor)
>                     cursor.movePosition(QtMoveEndOfWord, QtKeepAnchor)
>             self._vim.yankToInternal(cursor.selectedText())
>             cursor.removeSelectedText()
>             cursor.endEditBlock()
>             self.setTextCursor(cursor)
>             self._vim.resetCommand()
>             return True
> ```

I just added the test coverage for that :+1:  And expanded the doc on a slight difference in my implementation of the word end command "e" compared to how vim handles it.

**Comment by idealemu** [comment]
For me this is now at a state where I am moving my personal notes and world-building from markdown documents I used to edit in vim into novelWriter to have everything in one place. 
It might need more features (read more vim motions/commands) in the future, but it good enough for me to dog-food it for my personal novel writing :smile: 

Unless you have other points that I should address I'm willing to consider this PR ready.

(Now if novelWriter had the ability to have two editor windows open in a split I would be one happy novel writer !)

**Comment by vkbo** [comment]
> For me this is now at a state where I am moving my personal notes and world-building from markdown documents I used to edit in vim into novelWriter to have everything in one place. It might need more features (read more vim motions/commands) in the future, but it good enough for me to dog-food it for my personal novel writing 😄

😄 

> Unless you have other points that I should address I'm willing to consider this PR ready.

Nope. All good! I'm planning to do an RC1 release not this weekend, but next. So I'll merge it before this so it's included in RC1. After that it's usually two weeks of testing before I do the full 2.8 release.

> (Now if novelWriter had the ability to have two editor windows open in a split I would be one happy novel writer !)

Yeah, I'm missing that too sometimes when I'm writing but need to update something in my notes on the right side. I've thought about merging the editor and viewer classes into one and instead make it a edit/view mode in one panel. That way, you can switch either of them between edit and view as needed.

It's non-trivial but it may be worth the effort.
